### ISSUE TITLE
Wrong diff against specific revision for non-namespaced objects when using `application.resourceTrackingMethod: annotation`

### ISSUE THREAD

**Issue by jojand** [argoproj/argo-cd#9773]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

The tracking annotation `argocd.argoproj.io/tracking-id` contains namespace for non-namespaced objects. 
Running diff against specific revision (`argocd app diff --revision`) indicates deleting the namespace 
from the tracking annotation even if no change is going to happen.

**To Reproduce**

* have latest argocd-cli installed
* create a kind cluster (or any other test cluster alternative)
```bash
kind create cluster
```

* install Argo CD
```bash
kubectl create namespace argocd
kubectl apply -n argocd -f https://raw.githubusercontent.com/argoproj/argo-cd/v2.4.2/manifests/install.yaml
```

* patch `argocd-cm` to use `application.resourceTrackingMethod: annotation`
```bash
kubectl patch configmap/argocd-cm \
  -n argocd \
  --type merge \
  -p '{"data":{"application.resourceTrackingMethod":"annotation"}}'
```

* get initial password
```bash
kubectl -n argocd get secret argocd-initial-admin-secret -o jsonpath="{.data.password}" | base64 -d; echo
```

* apply simple test application (It contains ClusterRole which is a non-namespaced kubernetes object.)
```bash
cat <<EOF | kubectl apply -f -
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-clusterrole
  namespace: argocd
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: default
  source:
    path: test-clusterrole
    repoURL: https://github.com/jojand/argocd-test.git
EOF
```

* port-forward
```bash
kubectl port-forward service/argocd-server -n argocd 8080:80
```

* sync the application in UI
* app is in sync, no diff indicated

* run argocd app diff against HEAD revision (the current synced version)
```bash
argocd app diff test-clusterrole --revision=HEAD
===== rbac.authorization.k8s.io/ClusterRole /test-clusterrole ======
5c5
<     argocd.argoproj.io/tracking-id: test-clusterrole:rbac.authorization.k8s.io/ClusterRole:default/test-clusterrole
---
>     argocd.argoproj.io/tracking-id: test-clusterrole:rbac.authorization.k8s.io/ClusterRole:/test-clusterrole
```
The diff in the tracking annotations is not expected.

**Expected behavior**

```bash
argocd app diff test-clusterrole --revision=HEAD
```
No diff indicated in annotations.

**Screenshots**

**Version**
```
argocd version
argocd: v2.4.2+c6d0c8b.dirty
  BuildDate: 2022-06-22T02:07:16Z
  GitCommit: c6d0c8baaa291cd68465acd7ad6bef58b2b6f942
  GitTreeState: dirty
  GoVersion: go1.18.3
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v2.4.2+c6d0c8b
  BuildDate: 2022-06-21T20:42:05Z
  GitCommit: c6d0c8baaa291cd68465acd7ad6bef58b2b6f942
  GitTreeState: clean
  GoVersion: go1.18.3
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v4.4.1 2021-11-11T23:36:27Z
  Helm Version: v3.8.1+g5cb9af4
  Kubectl Version: v0.23.1
  Jsonnet Version: v0.18.0
```

**Logs**

**Comment by jessesuen** [comment]
Is the problem only with the CLI? Does the UI show a diff?

**Comment by coufalja** [comment]
AFAIK yes, no diff in UI. The sad thing is that we are using CLI as part of our PR checks and the bug reduces the usability of such checks.

**Comment by purajit** [comment]
Big +1 to this, we're testing out Argo but onboarding and moving services onto it has a slower iteration speed.

I've noticed that once the in-repo configuration has the correct name/kind/namespace/apiVersion of each resource, I'm able to then use the CLI to get the correct diffs. But often the naming conventions and what not are what we're trying to get right in the first couple iterations

**Comment by raxod502-plaid** [comment]
We have a similar situation, diff in CLI but not UI, and for non-namespaced object only. However we get a diff like this (namespace changes rather than going away):

```diff
-    argocd.argoproj.io/tracking-id: some-cluster:rbac.authorization.k8s.io/ClusterRoleBinding:some-other-ns/some-object-name
+    argocd.argoproj.io/tracking-id: some-cluster:rbac.authorization.k8s.io/ClusterRoleBinding:kube-system/some-object-name
```

**Comment by raxod502-plaid** [comment]
If you're using an ApplicationSet you can work around this by setting `applicationset.metadata.namespace` to `argocd`, but then `applicationset.spec.template.metadata.namespace` to match `applicationset.spec.template.destination.namespace`. No clue why as there is no documentation for that second field in the CRD.

**Comment by llavaud** [comment]
> If you're using an ApplicationSet you can work around this by setting `applicationset.metadata.namespace` to `argocd`, but then `applicationset.spec.template.metadata.namespace` to match `applicationset.spec.template.destination.namespace`. No clue why this works as there is no documentation for that second field in the CRD.

I guess you mean `applicationset.spec.template.spec.destination.namespace` for the last field, but this workaround don't work for me :(

**Comment by raxod502-plaid** [comment]
Yep, you are right with that correction. I have no idea why this works for us, so I'm also unclear on why it wouldn't work in other cases.

**Comment by maxbrunet** [comment]
Found the issue and proposed a fix in #13924
